driver-sync/src/main/com/mongodb/client/internal/ClientSessionImpl.java

```diff
@@ -220,7 +224,7 @@ public <T> T withTransaction(final TransactionBody<T> transactionBody, final Tra
                 if (transactionState == TransactionState.IN) {
                     abortTransaction();
```

Comment:
It seems when one of the operations holds a lock, executing another operation on that lock in `withTransaction` might lead to a timeout breach.

For example: the insert command fails with the following error:
`Command failed with error 24 (LockTimeout): 'Unable to acquire IX lock on '{6504462206465692013: Database, 1892776188038304109}' within 5ms. opId: 11901851, op: conn653927, connId: 653927.' on server localhost:27017. The full response is {"errorLabels": ["TransientTransactionError"], "ok": 0.0, "errmsg": "Unable to acquire IX lock on '{6504462206465692013: Database, 1892776188038304109}' within 5ms. ``

The transaction does not seem to have started due to the lock being held. When this `MongoCommandException `occurs and the transaction is attempted to be aborted, a fresh `TimeoutContext` is created with a new timeout. Since the error is classified as a "TransientTransactionError," the loop continues to operate, executing `startTransaction` with a new fresh timeout one more time.

Because the timeout for `withTransaction` is created fresh for `startTransaction`, it leads to an infinite loop until the lock is released, which breaches the timeout expectation for `withTransaction` as specified in the spec: "If timeoutMS is set, drivers MUST apply it to the entire withTransaction call."

I was able to reproduce it in `test10CustomTestWithTransactionUsesASingleTimeout`.

Reply:
Fantastic - I totally overlooked that scenario. Have added a new regression test 👍